src/fknni/faiss/faiss.py

```diff
@@ -32,87 +31,156 @@ def __init__(
 
-        mask = ~np.isnan(X).any(axis=1)
-        X_non_missing = X[mask]
-
-        index = faiss.index_factory(X_non_missing.shape[1], self.index_factory)
-        index.metric_type = faiss.METRIC_L2 if self.metric == "l2" else faiss.METRIC_INNER_PRODUCT
-        index.train(X_non_missing)
-        index.add(X_non_missing)
-
-        self.index_ = index
-        self.global_fallbacks_ = (
-            np.nanmean(X, axis=0) if self.strategy in ["mean", "weighted"] else np.nanmedian(X, axis=0)
+        # Prepare fallback values, used to prefill the query vectors nan´s
+        # or as an imputation fallback if we can't build an index
+        global_fallbacks_ = (
+            np.nanmean(self.X_full, axis=0) if self.strategy in ["mean", "weighted"] else np.nanmedian(self.X_full, axis=0)
         )
 
-        return self
-
-    def transform(self, X: np.ndarray | pd.DataFrame) -> np.ndarray:
-        """Imputes missing values in the data using the fitted Faiss index.
-
-        Args:
-            X: Data with missing values to impute. Expected to be either a NumPy array or a pandas DataFrame.
+        # We will need to impute all features having nan´s
+        feature_indices_to_impute = [i for i in range(self.X_full.shape[1]) if np.isnan(self.X_full[:, i]).any()]
+
+        # Now impute iteratively
+        while feature_indices_to_impute:
+            feature_indices_being_imputed, training_indices, training_data, index = self._fit_train_imputer(feature_indices_to_impute)
+
+            # Use fallback data if we can't build an index and iterate again
+            # TODO: Should we warn the user when this happens?
+            if index is None:
+                self.X_full[::, feature_indices_being_imputed] =  global_fallbacks_[feature_indices_being_imputed]
+                continue
+
+            # Extract the features from X that was used to train FAISS, and compute the sparseness matrix
+            x_imputed = self.X_full[::, training_indices]
+            x_imputed_missing_mask = np.isnan(x_imputed)
+
+            # Iterate through the data to impute, ignoring already imputed rows
+            sample_missing_mask = []
+            for idx in np.where(x_imputed_missing_mask.any(axis=1))[0]:
+                # Extract the sample
+                sample_missing_mask = x_imputed_missing_mask[idx]
+                sample = x_imputed[idx]
+
+                # We need to prefill the query vector as FAISS doesn't accept nan´s
+                sample[sample_missing_mask] = global_fallbacks_[training_indices][sample_missing_mask]
+
+                # Call FAISS and retrieve data
+                distances, indices = index.search(sample.reshape(1, -1), self.n_neighbors)
+                valid_indices = indices[0][indices[0] >= 0] # Filter out negative indices because they are FAISS error codes
+                assert len(valid_indices) >=1  # TODO: Assert or warning (if at least 1 valid neighbor?)
```

Comment:
asserts are only for developers because they can get stripped away in builds. Always raise other error types (it's easy to create custom exceptions). We cannot exit with an error because that would effectively mean that this strategy could not impute ALL datasets, right? I think we have to implement a solution that always works but for super trashy datasets may just return bad results. Nothing we can do.